### PR TITLE
Fix OLMo 2 attention input fork state

### LINKED ISSUE #1687: [Bug Report] OLMo 2: use_split_qkv_input / use_attn_in corrupt the forward pass, non-deterministically, across repeated runs
**Describe the bug**

On OLMo 2 bridges, enabling `set_use_split_qkv_input(True)` (or `set_use_attn_in(True)`) makes the forward pass **wrong and non-reproducible**: repeating the *same* `run_with_cache` call on the same tokens returns different logits each time, drifting further from the correct values with each run. The error is the same order of magnitude as the signal, so results look plausible while being meaningless, and re-running does not reproduce them.

Measured on `hf-internal-testing/tiny-random-Olmo2ForCausalLM` (compatibility mode, fresh bridge, setter called before any forward), comparing against a reference bridge with the flag off — reference logit scale is **25.68**:

| run | max abs logit delta vs reference | `blocks.1.attn.hook_q_input − blocks.1.hook_resid_pre` |
|---|---|---|
| 1 | 30.49 | 0.0000 |
| 2 | 29.33 | 56.03 |
| 3 | 24.21 | 60.91 |
| 4 | 26.92 | 63.14 |

A `distilgpt2` control run through the identical sequence is flat at `1.81e-05` on all runs. The `hook_q_input` column shows the split-QKV residual copy diverging from `hook_resid_pre`, which it should equal.

**Code example**

```python
import torch
from transformer_lens.model_bridge import TransformerBridge

M = "hf-internal-testing/tiny-random-Olmo2ForCausalLM"
toks = torch.tensor([[1, 2, 3, 4, 5]])

reference = TransformerBridge.boot_transformers(M, device="cpu")
reference.enable_compatibility_mode()
expected = reference(toks, return_type="logits").clone()

bridge = TransformerBridge.boot_transformers(M, device="cpu")
bridge.enable_compatibility_mode()
bridge.set_use_split_qkv_input(True)

for i in range(4):
    bridge.run_with_cache(toks)
    delta = (bridge(toks, return_type="logits") - expected).abs().max().item()
    print(i, delta)   # 30.49, 29.33, 24.21, 26.92 — different every run
```

**System Info**

Installed from source on `dev-4.x`; macOS arm64, CPU, fp32, Python 3.12, transformers 5.13.0. Not architecture-specific to the tiny fixture — it is the OLMo 2 adapter path.

**Expected behaviour & fix pointer**

Two invariants are violated and both should hold:

1. **Determinism** – repeating an identical forward on an unchanged model must return identical activations and logits. State must not accumulate across forwards.
2. **Correctness of the split fork** – with `use_split_qkv_input`, each of `hook_q_input` / `hook_k_input` / `hook_v_input` must equal the residual entering the block (`hook_resid_pre`) before any patching, as it does on pre-norm architectures (verified `0.0` on gpt2/distilgpt2).

Start at `_propagate_attention_flag` (`transformer_lens/model_bridge/transformer_bridge.py:3621`, called from `set_use_split_qkv_input` at `:3608` and `set_use_attn_in` at `:3623`) and at the OLMo 2 adapter's attention wiring. The drift pattern points at per-forward state that is written but never reset between runs. Note OLMo 2 is post-norm, so the fork point differs from the pre-norm families where this path is exercised by tests.

Acceptance:
- [ ] Repeating an identical `run_with_cache` on an OLMo 2 bridge with `use_split_qkv_input=True` returns identical logits and activations across runs
- [ ] `hook_q_input` / `hook_k_input` / `hook_v_input` equal `hook_resid_pre` on a fresh run, or the setter refuses with `NotImplementedError`
- [ ] Same for `set_use_attn_in(True)`, which shows the same signature
- [ ] Regression test under `tests/unit/model_bridge/` (or `tests/integration/model_bridge/`) that runs the same input **at least twice** and asserts equality — a single-run test cannot catch this
- [ ] `make unit-test` passes
- [ ] `uv run mypy .` passes

**Additional context**

### Checklist
- [x] I have checked that there is no similar issue in the repo (**required**)

### PR DESCRIPTION
**Pull request by Excelius-Wang**
# Description

OLMo 2 uses post-norm blocks, so `ln1` runs after attention. `BlockBridge` was wiring every `ln1` as the pre-attention capture for `use_split_qkv_input` and `use_attn_in`. The captured post-attention tensor then leaked into the next forward and replaced the current residual input, causing logits and activations to drift across identical runs.

Skip the pre-`ln1` capture for blocks already marked as reading the residual directly, and add regression coverage for both attention input fork modes. The tests verify that the fork input matches `hook_resid_pre`, outputs remain baseline-equivalent, and repeated forwards are bitwise deterministic.

Fixes #1687

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Testing

- `pytest tests/integration/model_bridge/test_olmo2_hook_semantics.py -q` — 9 passed
- `pytest tests/unit/model_bridge/supported_architectures/test_olmo2_adapter.py tests/unit/model_bridge/supported_architectures/test_olmo3_adapter.py tests/unit/model_bridge/supported_architectures/test_olmo_adapter.py tests/unit/model_bridge/supported_architectures/test_olmo_hybrid_adapter.py tests/unit/model_bridge/supported_architectures/test_flex_olmo_adapter.py -q` — 62 passed
- `pytest tests/unit/model_bridge/compatibility/test_split_qkv.py -q` — 7 passed
- `make unit-test` — 5029 passed, 35 skipped, 44 deselected, 8 xfailed
- `uv run mypy .`
- `make check-format`

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Review by koriyoshi2041 — COMMENTED** [review]
#1699 already merged the regression fix for #1687. Its block-entry clear and `finally` cleanup make the attention fork fall back to the current `resid_pre` before attention runs, while also bounding the lifetime of the transient capture for every architecture.

This alternative is therefore redundant after `a5894f5`, and `mlp_reads_resid_directly` is a less direct proxy for whether `ln1` is pre- or post-attention. The added OLMo 2 assertions are useful, but the implementation should not land on top of the merged lifetime fix. I would close this unless the tests demonstrate a failure on current `main`.

**Comment by Excelius-Wang** [comment]
Thanks for catching this. My branch was based on c444ae1, before #1699 merged as a5894f5.

I agree that the implementation is now redundant, and that mlp_reads_resid_directly is not the right architectural signal for determining whether ln1 is pre- or post-attention. #1699 already covers both attention-input fork modes and the cross-forward state lifetime.

I’ll verify the additional OLMo 2 assertions against the current dev branch without this implementation change. If they pass, I’ll close this PR; if they expose any remaining gap, I’ll propose the coverage separately as a tests-only change.

**Comment by Excelius-Wang** [comment]
Verified against current dev at 5b23538 with only the additional OLMo 2 assertions applied and no implementation changes. Both split-QKV and attn-in cases pass, including the resid_pre input, baseline-equivalence, and repeated-forward determinism checks.

This confirms that #1699 fully covers the regression, so I’m closing this PR as superseded. Thanks again for the review.

**Comment by jlarson4** [comment]
@Excelius-Wang Thank you for taking the time to do the do-diligence work on existing fix, and thank you @koriyoshi2041 for letting him know the fix was already implemented. Apologies for forgetting to close the issue earlier when I merged that PR.
